File: FreeBuildJ/net/cemetech/sfgp/freebuild/drivers/Drivers.java
package net.cemetech.sfgp.freebuild.drivers;

import java.util.*;

public class Drivers {

	private static Drivers instance = null;
	
	private AbstractDriver mainDriver;
	private HashMap<String,Class<? extends AbstractDriver>> knownDrivers;
	
	protected Drivers(){
		mainDriver = null;
		knownDrivers = new HashMap<String,Class<? extends AbstractDriver>>();
	}
	
	public static Drivers manager(){
		return (instance == null) ? (instance = new Drivers()) : instance; 
	}
	
	public void setMainDriver(AbstractDriver nd){
		mainDriver = nd;
	}
	
	public AbstractDriver getMainDriver(){
		return mainDriver;
	}
	
	public void clearMainDriver(){
		mainDriver = null;
	}
	
	public Class<? extends AbstractDriver> getNamedDriver(String name){
		return knownDrivers.get(name);
	}
	
	public Class<? extends AbstractDriver> registerNamedDriver(String name, Class<? extends AbstractDriver> newDriver){
		return knownDrivers.put(name,newDriver);
	}
	
	public Class<? extends AbstractDriver> unregisterDriver(String name){
		return knownDrivers.remove(name);
	}
	
}

File: FreeBuildJ/net/cemetech/sfgp/freebuild/drivers/Drivers.scala
package net.cemetech.sfgp.freebuild.drivers;

import scala.collection._

object Drivers {
	
	protected var mainDriver:AbstractDriver = null
	protected var knownDrivers:Map[String, Class[_ <:  AbstractDriver]] = Map[String, Class[_ <: AbstractDriver]]()
		
	def setMainDriver(nd:AbstractDriver):Unit = { mainDriver = nd }
	def getMainDriver():AbstractDriver = { mainDriver }
	def clearMainDriver():Unit = { mainDriver = null }
	
	def getNamedDriver(name:String):Class[_ <: AbstractDriver] = { knownDrivers.getOrElse(name,null) }
	
	def registerNamedDriver(name:String, newDriver:Class[_ <: AbstractDriver]):Class[_ <: AbstractDriver] = { 
		val current = getNamedDriver(name)
		knownDrivers = knownDrivers + (name -> newDriver)
		current
	}
	
	def unregisterDriver(name:String) = {
		val current = getNamedDriver(name)
		knownDrivers = knownDrivers - name
		current
	}
	
}
